exporter/datadogexporter/internal/logs/translator.go

```diff
@@ -0,0 +1,172 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logs // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter/internal/logs"
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/otlp/model/attributes"
+	"github.com/DataDog/datadog-agent/pkg/otlp/model/source"
+	"github.com/DataDog/datadog-api-client-go/v2/api/datadog"
+	"github.com/DataDog/datadog-api-client-go/v2/api/datadogV2"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	conventions "go.opentelemetry.io/collector/semconv/v1.6.1"
+)
+
+const (
+	namespace          = "otel"
+	otelTraceID        = namespace + ".trace_id"
+	otelSpandID        = namespace + ".span_id"
+	otelSeverityNumber = namespace + ".serverity_number"
+	otelSeverityText   = namespace + ".serverity_text"
+	otelTimestamp      = namespace + ".timestamp"
+
+	ddStatus    = "status"
+	ddTraceID   = "dd.trace_id"
+	ddSpanID    = "dd.span_id"
+	ddTimestamp = "@timestamp"
+)
+
+const (
+	logLevelTrace = "trace"
+	logLevelDebug = "debug"
+	logLevelInfo  = "info"
+	logLevelWarn  = "warn"
+	logLevelError = "error"
+	logLevelFatal = "fatal"
+)
+
+// Transform is responsible to convert LogRecord to datadog format
+func Transform(lr plog.LogRecord, res pcommon.Resource) datadogV2.HTTPLogItem {
+	hostName, serviceName := extractHostNameService(res.Attributes(), lr.Attributes())
+
+	l := datadogV2.HTTPLogItem{
+		AdditionalProperties: make(map[string]string),
+	}
+	if hostName != "" {
+		l.Hostname = datadog.PtrString(hostName)
+	}
+
+	if serviceName != "" {
+		l.Service = datadog.PtrString(serviceName)
+	}
+
+	// we need to set log attributes as AdditionalProperties
+	lr.Attributes().Range(func(k string, v pcommon.Value) bool {
+		l.AdditionalProperties[k] = v.AsString()
+		return true
+	})
+	if !lr.TraceID().IsEmpty() {
+		l.AdditionalProperties[ddTraceID] = convertTraceID(lr.TraceID().HexString())
+		l.AdditionalProperties[otelTraceID] = lr.TraceID().HexString()
+	}
+	if !lr.SpanID().IsEmpty() {
+		l.AdditionalProperties[ddSpanID] = convertTraceID(lr.SpanID().HexString())
+		l.AdditionalProperties[otelSpandID] = lr.SpanID().HexString()
+	}
+	var status string
+
+	if lr.SeverityText() != "" {
+		status = lr.SeverityText()
+		l.AdditionalProperties[otelSeverityText] = lr.SeverityText()
+	} else {
+		status = deriveStatus(lr.SeverityNumber())
+	}
+	l.AdditionalProperties[ddStatus] = status
```

Comment:
It looks like right now, if the severity text is present, we use that as the status. Only if the severity text is not present do we derive the status from the severity number.

Is this intentional? I proposed the opposite, because the severity number is bounded/finite whereas the severity text is arbitrary/unbounded. That is, I think we should use the severity number as the source of truth for the status of the log. If there is no severity number, I think we could try to make a best-effort conversion of the severity text to a status. However, I don't think we can just use the severity text as-is in any case, as it might not correspond to a valid Datadog status level.

Reply:
Most likely , we have to use [severity_parser](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/af4e1ef5c3109d4c9dc4700a0b2c2bcb17264e5e/pkg/stanza/operator/helper/severity.go#L43) to get severity , that would mean using either should be ok .

Reply:
I'm not familiar with severity_parser, is this a different component that the customer can include in the pipeline before the exporter? If so I don't think we should assume that the customer is using it, I think the logic in the exporter should work regardless of what other parsing the customer is performing in advance.

Reply:
we need to some parsing to convert logs to otel log record. The adapters or sdks for each language to emit logs in otel log record format are still in experimental phase.

Reply:
I have reordered the condition.

Reply:
Per my other comment, given that the log levels are different from what I thought and what I included in the original RFC, I think it might actually make more sense to prioritize severity text 🙁 sorry for the confusion.